### DISCUSSION TITLE
Prettier Options compatibility

### DISCUSSION THREAD

**Discussion by MichaReiser** [rome/tools#2465] [General]
The goal of this discussion is to come to an agreement on what JavaScript [Prettier options](https://prettier.io/docs/en/options.html) Rome should support and what values it should use as defaults. The discussion will also serve as a reference for our reasoning in the future. 

Rome supports the following Prettier options today:
* [Print width](https://prettier.io/docs/en/options.html#print-width)
* [Tab width](https://prettier.io/docs/en/options.html#tab-width)
* [Quotes](https://prettier.io/docs/en/options.html#quotes)

These options aren't part of this discussion except if Rome and Prettier use different defaults. 

Supporting fewer options is desirable from a technical position because it reduces the number of possible option combinations, which reduces the overall complexity. The main non-technical benefit is that it removes debates about what the "right" formatter configuration is, one of the main goals of Rome's formatter. You can read more in Prettier's excellent [philosophy documentation](https://prettier.io/docs/en/option-philosophy.html).

This discussion is related to the [Formatter options](https://github.com/rome/tools/discussions/2318) discussion but is focused on inheriting options from Prettier.

**Comment by MichaReiser** [comment]
## [Tabs](https://prettier.io/docs/en/options.html#tabs)

Both Rome and Prettier support the `Tabs` options but Prettier defaults to `spaces` whereas Rome defaults to `tabs`. 

Rome defaults to `tabs` because they are more accessible. See this thread for a past discussion https://github.com/rome/tools/issues/425

### Proposal

Keep tabs as the default. Default to `spaces` if migrating a project from prettier.

**Reply by ematipico** [reply]
I agree with the proposal

**Reply by vjeux** [reply]
Since you have a way to parse how many prettier configs have each, would be interesting to know how many have switched to tabs.

**Reply by MichaReiser** [reply]
There are about [200k usages](https://github.com/search?q=useTabs+filename%3A.prettierrc&type=code) of the `useTabs` setting in `prettier.rc` files. But this includes `useTabs: false` and `useTabs: true`. That's about 20% of all config files (minus all the project that don't have a configuration file).

**Reply by scarf005** [reply]
I suggest rome to default to use spaces.
 
### rationale
it would safe to assume large portion of rome users would expect its defaults to use `spaces`, since search from https://cs.github.com shows that 89.7% of prettier config uses spaces, and 1/3 of all prettierconfig explicitly use spaces.

| Options     | explicit `useTab: true` | explicit `useTab: false` | implicit `useTab: false` | Total      |
| ----------- | ----------------------- | ------------------------ | ------------------------ | ---------- |
| Usages      | [6.8k][6.8k]            | [25k][25k]               | 34.2k (estimated)        | [66k][66k] |
| Percentages | 10.3%                   | 37.8%                    | 51.9% (estimated)        | 100%       |

[6.8k]: https://cs.github.com/?scopeName=All+repos&scope=&q=%2F.%3FuseTabs.%3F%3A+.%3Ftrue.%3F%2F+path%3A.prettierrc*
[25k]: https://cs.github.com/?scopeName=All+repos&scope=&q=%2F.%3FuseTabs.%3F%3A+.%3Ffalse.%3F%2F+path%3A.prettierrc*
[66k]: https://cs.github.com/?scopeName=All+repos&scope=&q=path%3A.prettierrc*

and in current `rome.json` usage, it shows that around 59% of usages explicitly uses `indentStyle: space` .
| Options     | explicit `indentStyle: space` | explicit `indentStyle: tab` | implicit `indentStyle: tab` | Total    |
| ----------- | ----------------------------- | --------------------------- | --------------------------- | -------- |
| Usages      | [49][49]                      | [9][9]                      | 25                          | [83][83] |
| Percentages | 59.04%                        | 10.84%                      | 30.12%                      | 100%     |

[49]: https://cs.github.com/?scopeName=All+repos&scope=&q=%2F.%3FindentStyle.%3F%3A+.%3Fspace.%3F%2F+path%3A%2F^rome.json%24%2F
[9]: https://cs.github.com/?scopeName=All+repos&scope=&q=%2F.%3FindentStyle.%3F%3A+.%3Ftab.%3F%2F+path%3A%2F^rome.json%24%2F
[83]: https://cs.github.com/?scopeName=All+repos&scope=&q=+path%3A%2F^rome.json%24%2F

**Comment by MichaReiser** [comment]
## [Quote Props](https://prettier.io/docs/en/options.html#quote-props)

The quote props option specifies if the formatter should remove quotes around props that aren't syntactically necessary. Rome doesn't support this option today. 

The reason why quote props is configurable in Prettier is that quoted props prevent minifiers from mangling that property name. 

### Proposal

Don’t add support for quote props and hint users towards `rome-ignore` to disable Rome from stripping the quotes. This assumes, that disabling the behaviour is only needed in rare cases.

### Note

Implementing the removal of quote props is tracked in https://github.com/rome/tools/issues/2405

**Reply by ematipico** [reply]
This is a though one. We don't know what would be the impact of this one on minifiers, in case we decide to not support it. If a minifier messes up, finding the issue might take a bit of time on the developer.

**Reply by MichaReiser** [reply]
Prettier's default is to remove the quotes. Finding the issue might be difficult because it only happens in production but you get a source line and roughly know which property it is. The main challenge is if there are multiple places where the object gets instantiated.

I would would recommend people to change their minifier configuration if they face this problem all over there code base or use rome ignore to suppress specific cases that are problematic

**Reply by Conaclos** [reply]
Including quotes around a property is often a deliberate choice. In my opinion it could be preferable to default to "preserve". The option could be removed and rome could preserve every quote around properties.

**Reply by MichaReiser** [reply]
@Conaclos you're saying that adding quotes is a deliberate choice. When do you deliberately add quotes around props?

**Reply by Conaclos** [reply]
@MichaReiser 
In JavaScript, it is common to use an object as a dictionary. To do that we use a bracket notation:

```js
dict[key] = val
```

[I sometimes use](https://github.com/bare-ts/tools/blob/main/src/ast/bare-ast.ts#L288) "statically-defined" maps:

```js
const DICT  = {
    "key": val,
}
``` 
In this case, I deliberately quote the properties, to forbid minifiers and compilers (such as Google closure) to mangle these properties. **This could also be used for types serialized in and deserialized from JSON**.

I could even like to have a type checker that enforce bracket notations with quoted properties and dot notations with unquoted properties.

```ts
interface SerializablePerson {
    "name": string
}
interface ManglablePerson {
    name: string
}

declare let p1: SerializablePerson
declare let p2: ManglablePerson

p1 = p2 // error
p2 = p1 // error
p1.name = "Seldon" // error
p2["name"] = "Seldon" // error

p1["name"] = "Seldon" // ok
p2.name = "Seldon" // ok
for (const x in p2) {
    if (Object.hasOwnProperty.call(p1, x)) {
        p1[x] // ok because dynamic access
    }
}

JSON.stringify(p1) // ok
JSON.stringify(p2) // warning
```

**Reply by Conaclos** [reply]
If you assume that properties are not mangled, it is fine to remove quotes.

**Reply by MichaReiser** [reply]
> In this case, I deliberately quote the properties, to forbid minifiers and compilers (such as Google closure) to mangle these properties.

How frequent is this use case in your code-base? Would using a `rome-disable: formatter Don't remove quotes to avoid minifiers from mangle the properties`. 

The main downside I see of using `rome-disable`/`rome-ignore` is that this also disables all other formatting which probably isn't what you want. 

This may be enough reason to add a `preserve` option but default to removing them if not necessary (same as prettier).

**Reply by Conaclos** [reply]
> How frequent is this use case in your code-base?

Basically every time I need a static map. A comment could be okish. I am not a big fan of comments because it is tool-specific.

> The main downside I see of using rome-disable/rome-ignore is that this also disables all other formatting which probably isn't what you want.

+1

> This may be enough reason to add a preserve option but default to removing them if not necessary (same as prettier).

Is there any reason for removing them (apart to be prettier-compliant)? I think it is pretty rare that people quote properties without deliberating to?

**Reply by MichaReiser** [reply]
> Is there any reason for removing them (apart to be prettier-compliant)? I think it is pretty rare that people quote properties without deliberating to?

Prettier removes the quotes by default and I would draw the conclusion from a quick code search that people are mostly OK with that:

* [consistent](https://github.com/search?q=filename%3A.prettierrc+quoteProps+consistent&type=code) 9K results
* [preserve](https://github.com/search?q=filename%3A.prettierrc+quoteProps+preserve&type=code) 17K

Out of roughly [1M](https://github.com/search?q=filename%3A.prettierrc&type=code) `.prettierrc` files. 

I'm now a bit undecided for which options Rome should support

* `consistent`: Makes the most sense for the use case that you mentioned where you either want all props to be quoted or none of them. 
* `preserve`: Seems to be more widely used and allows more granular control. But I'm failing to see a use case where you want to disable mangling of a single property. Shouldn't it be all or none?

**Reply by Conaclos** [reply]
@MichaReiser 
In fact "consistent" does not solve the issue because it quotes all properties if and only at least one property **requires** quotes. In my previous example this option could then remove the quotes. "consistent" is purely stylistic.

I think _rome_ may drop the "consistent" option and either (1) provide an option with "as-needed" / "preserve" (default to one of them), or (2) no option and use the "preserve" behavior (if users complain about that, an option for "as-needed" could be added).

**Reply by MichaReiser** [reply]
@Conaclos I can see how `preserve` is the most flexible. But my understanding from your "object as a dictionary" use case is that you would want to quote all props in that case, which is why `consistent` is a better fit in my view. 

And thanks for running through your use case with me. This is extremely valuable.

**Reply by Conaclos** [reply]
@MichaReiser 
The behavior you describe is a mix of "consistent" and "preserve".
Again "consistent" removes quotes in cases where there are not **needed**.

As an example, this code
```js
const DICT = {
  "key": 1,
};
```
[is formatted](https://prettier.io/playground/#N4Igxg9gdgLgprEAuc0DOMAEARAkgYQBVMBeTYAHSk0wpAGs4BPOpTARgBoqBfAbhCcQEAA4wAlumSgAhgCc5EAO4AFeQjTIQMgDZKZTTUIBGcmWEYwAyjIC2cADLiocZADNdaOCbMW41kXNnAHNkGDkAV28QOFtjOAATBMSHGShgiJlguAAxCDlbGRgJdK0ZCJgIQRAACxhbHQB1GvF4NECwOCsNVvEAN1amLTA0IxBnLzkYFTNgwvdPaIArNAAPKxCdOABFCIh4BZ0vIUC5Sa1jGXidapE5ZxhG8QSYGuQADgAGE8UvRrMRFo7nBJn1XEIAI57eAzUSaFCQKBocQYBAwapyOBQ8SYmZZeZIDxHaJeWziMKREmbHbQ1yExZCGBXJ4vN5IABMjLM4h0IXwEFsBJiaAArNUIl5CFd4UTjiA+lFcFBkrArGB7mIAILKqwwJhbQ5eHg8IA) to the following code under "consistent":
```js
const DICT = {
  key: 1,
};
```
because quotes are not "needed" for this property.

Thanks for your attention :) Have a nice evening!

**Reply by Conaclos** [reply]
Note: I am also ok with this mix because it looks more consistent. However it is not possible to reproduce that in prettier. I do not know whether it is an issue...

**Reply by MichaReiser** [reply]
Oh, I understood that "consistent" would preserve the quotes if at least one prop uses quotes. But I guess, this doesn't make sense. 

Agree, Rome should then support "preserve" to cover the minifier use case and "as-needed" to remove them (default).

**Reply by vjeux** [reply]
The only reason we have this one is for Google Closure compatibility. In practice Google is the main group of people using Closure advanced mode that requires this distinction.

The amp project is using prettier and from what I’ve been told many more projects within Google are using it after this option was added.

main issue where this was discussed: https://github.com/prettier/prettier/issues/4327
Pr that implenented it: https://github.com/prettier/prettier/pull/5934

**Reply by ematipico** [reply]
Thank you everyone for the constructive conversion. It seems that this is an option that would like to maintain, so we will start working on it. 

As mention by @MichaReiser , we would need to have a default behaviour that mimics `"as-needed"` from Prettier, and then have a way to opt-in the `"preserve"` Prettier option.

That being said, default Rome's behaviour will be "as-needed" and then we can have an option called `--preserve-quotes` /`preserve_quotes` to opt-in the "preserve" one.

**Comment by MichaReiser** [comment]
## [Semicolons](https://prettier.io/docs/en/options.html#semicolons)

Prettier supports the semicolons option that specifies if semicolons should be inserted at the end of statements. 

## Proposal

Don't add support for it. My research shows that the option isn't widely used:
* [1K](https://github.com/search?l=&q=noSemi+filename%3A.prettierrc&type=code) appearances of the `noSemi` option
* [2.3M `.prettierrc`](https://github.com/search?q=filename%3A.prettierrc&type=code) files existing on GitHub (which doesn't account for everyone using the defaults). 


Edit: 
The configuration option is `semi: false` and not `noSemi`. GitHub search shows around 230k usages of [`semi: false`](https://github.com/search?p=6&q=%22semi%3A+false%22+filename%3A.prettierrc&type=Code)

**Reply by ematipico** [reply]
Agree with the proposal. Not using semicolons requires also [particular lint rules](https://github.com/standard/standard-engine) to avoid errors.

**Reply by Conaclos** [reply]
Some famous formatter such as [StandardJS](https://github.com/standard/standard) removes semicolons.

**Reply by MichaReiser** [reply]
> Some famous formatter such as [StandardJS](https://github.com/standard/standard) removes semicolons.

Semicolons are a question of preference and `StandardJS`s preference is to go without semicolons and they are intentionally opinionated about it:

> No. The whole point of standard is to save you time by avoiding [bikeshedding](https://docs.freebsd.org/en/books/faq/#bikeshed-painting) about code style. There are lots of debates online about tabs vs. spaces, etc. that will never be resolved. These debates just distract from getting stuff done.

The main question is if it's reasonable for Rome to be as opinionated as well and default to semicolons. StandardJS remains a valid alternative for everyone who prefers the StandardJs code style (from which Rome probably diverges quite a bit). 

@Conaclos is your preference to not use semicolons? Would you still adopt Rome even if it ships without the option to remove semicolons but otherwise fulfils all your needs?

**Reply by Conaclos** [reply]
I must admit that I never liked semicolons. Most of people I met defended semicolons until they play with a language without semicolons, and then they realize that their code is more readable without semicolons. Arguments about safety no longer hold in a world with formatters, linters, bundlers, and/or compilers.

However, I found workarounds to reduce the noise introduced by semicolons (I customized my code editor theme to make them less visible). Thus, whether a tool is worth to accept the annoyance of semicolons I could tolerate their presence ^^

**Reply by vjeux** [reply]
Note that the no-semi option in prettier doesn’t just remove semi-colons at end of statements. There’s a complex piece of logic to know when the last character printed would be a ) of ] and therefore not putting a semi would change the meaning. In this case it puts the semi colon at the beginning of the next line.

I personally don’t use no-semi but it caused so much noise not having it that it was better for its adoption than not.

**Reply by vjeux** [reply]
Code for the logic: https://github.com/prettier/prettier/blob/199a1bb38929c72861ead3cf44978a895eba65bb/src/language-js/print/statement.js#L89

**Reply by ematipico** [reply]
Thank you everyone! It totally makes sense to have this option, but I am not sure if Rome is ready to have it **now**. As we know, the usage of semicolons is crucial in some parts of the code.

In order to achieve that, a complex logic inside the formatter (like prettier does), or inside a linter (like StandardJS does) are crucial for this to land.

As Rome aims to be a formatter **and** an analyzer, the team first has ask themselves the question: who's in charge of checking the correct presence of semicolons inside a semicolon-free style?

Once we answer this question, we can then move forward as we see fit. As for now, I think we should put on hold the discussion

**Reply by michaelhays** [reply]
As a correction to the research in the proposal, the [prettier configuration option](https://prettier.io/docs/en/options.html#semicolons) is actually `"semi"`, not `"noSemi"`.

In reality, there are [over 300K instances](https://github.com/search?p=5&q=%22semi%3A+false%22+filename%3A.prettierrc&type=Code) of `"semi": false`, which seems significant.

(As a personal aside, I really hate semicolons, and not having the option to remove them in formatting would be hard to swallow. But I hate bikeshedding more, and would understand if you ultimately made the decision to forgo adding the option.)

**Reply by michaelhays** [reply]
@MichaReiser Do you mind updating your analysis in your original post to use `semi` instead of `noSemi`? (See my comment above.) I feel like the prevalence of the option changes this discussion quite a bit.

I'd really love to adopt to Rome right now -- currently struggling with some awfully slow performance of ESLint and Prettier in a complex codebase -- but honestly the lack of an option to remove semicolons makes it difficult to rationalize, considering the massive amount of changes it would make to my existing codebase. This is of course just my personal experience, and I understand that adding support for the option requires some complex logic, but I imagine it's a piece of feedback you'll start receiving a lot.

**Reply by MichaReiser** [reply]
Sorry, I missed your first comment. Thanks for pointing out the correct name of the configuration option. I updated my post with these numbers (GitHub search yields very different results every time I refresh the page). 

There's no strong reason for Rome not adding support for `noSemi` and it also doesn't seem too complicated. Our reasoning on what options to implement is mainly around prioritisation and `noSemi`doesn't make the cut at the time being. I would support a community contribution (from quickly scanning the code, it seems that it is only concerI recommendements, not making it an over complicated change). 

I recommend that you open an issue to track interest in the `in noSemi` option which, in turn, can help to find a contributor interested implementing the change. Let me know if you're interested in contributing and I can support you in making the change.

**Reply by michaelhays** [reply]
Awesome, thanks for updating that! And same, I got anywhere from 100k-400k from the GitHub search, not sure what's going on there :sweat_smile: Looks like it increases every time I refresh the page and stabilizes at 422k? But I'm not sure if that's accurate.

Yep, definitely get the prioritization. I just opened [this issue](https://github.com/rome/tools/issues/3535) to track interest. I'd love to contribute, but admittedly will be slammed for the next few months, and may have a hard time finding more than a few hours to contribute before then. I'll definitely try to find some time soon though and take a deeper dive into the implementation!

**Reply by mrm007** [reply]
> In reality, there are [over 300K instances](https://github.com/search?p=5&q=%22semi%3A+false%22+filename%3A.prettierrc&type=Code) of `"semi": false`, which seems significant.

I'm pretty sure that number should be taken with a grain of salt. You'll notice that some files contain multiple hits, which significantly inflates the total number of hits (here's a [random example](https://github.com/wsx-template-projects/template-other/blob/6079d12efcc1b6e13d09e1de1b3f886f43738186/other/template-vue3-app-h5/.prettierrc.js) with 3 matches).

If the numbers reported by GitHub Search are to be believed, then we should add up the total of:
- results for JavaScript (between 15k-30k)
- results for JSON (between 15k-30k)
- results for YAML (between 3k-6k)

Total: 30k-70k, depending on the phase of the moon. An order of magnitude lower than the quoted 300k.

[GitHub Code Search](https://cs.github.com/?scopeName=All+repos&scope=&q=%2F.%3Fsemi.%3F%3A+false%2F+path%3A.prettierrc*) reports around 23k total, with one hit per file.

**Reply by bilalq** [reply]
The number also undercounts shared configs and private repos. For example, we have dozens of repos using prettier without semicolons from a single shared prettier config. The `"semi": false` pattern would only appear once. I'm sure many other orgs experience something similar.

**Reply by bilalq** [reply]
Also another 21k or so from package.json usages of the semi: false config:

https://github.com/search?p=5&q=%22semi%3A+false%22+filename%3Apackage.json&type=Code

**Reply by jeysal** [reply]
IIRC in the static site generator space, Gatsby etc., as well as with MDX users, `semi: false` is quite popular

**Reply by MichaReiser** [reply]
Thank you all for the number crunching! The support for `semi: false` is tracked in #3535. We see it as a valuable option that supports the setup of many people in the community, but it is't a high priority. Leave a comment on the issue if you're interested in implementing the feature so that we can help you.

**Comment by MichaReiser** [comment]
## [Trailing Commas](https://prettier.io/docs/en/options.html#trailing-commas)

The trailing commas option configures where to insert trailing commas. It supports three different configurations:

* `none`: Removes all trailing commas
* `es5`: Adds trailing commas in places permitted by ES5
* `all`: Add trailing commas everywhere

The reason Prettier added this option is for backward compatibility because trailing commas have only been supported in a few places prior to ES2017 and TypeScript 2.7. Thus, it would have been necessary to transform the source code if targeting a runtime that doesn't support ES2017. 

* Node supports ES2017 trailing commas since version 8.2 which was released in 2017 (now 17.9)
* All major browsers support ES 2017 trailing commas since mid-2017 [source (at the very end)](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Trailing_commas)

## Proposal

Don’t support the trailing comma option because all modern engines support them today and the sole reason the option was introduced by Prettier is for compatibility’s sake. Prettier even changed the default from `none` to `es5` in their `v2` release.

## Alternatives

Implement the option and either:

- Support `es5` and `all`
- Support `none`, `es5`, and `all`

**Reply by leops** [reply]
We may want to consider supporting `none` and `all` as a third alternative: currently the trailing commas are already conditionally inserted anyway (they only get printed if the surrounding group breaks) and it would be relatively easy to support `none` by skipping all trailing commas unconditionally, but supporting `es5` would mean having to track a language version for each position where we insert a trailing comma

**Reply by MichaReiser** [reply]
My understanding is that our current behaviour matches Prettier's `all` by only inserting a comma if a group breaks. 

I don't think we need to track the language version for `es5`. It should be possible to not generate the trailing comma token for function arguments if the `traling_coma` option is set to `es5`. 

Is your view that Rome should add support for the trailing comma option? If so, what's your reasoning?

**Reply by leops** [reply]
If the only difference between `es5` and `all` if in function arguments I guess it's alright to support both, but otherwise I think supporting `es5` would just add more complexity than necessary.
Overall I don't think we should implement support for the trailing commas option, to me the only reason we would want to implement it would be specifically to be compatible with Prettier: if we're not aiming for 100% compatibility with Prettier only implementing the `all` behavior seems like the better option.

**Reply by ematipico** [reply]
I agree with the proposal. Unless we aim to use the formatter on IE11, I don't see the value of the additional complexity.

**Reply by vjeux** [reply]
It’s amazing that all the mainstream parsers now support trailing commas!

**Reply by Th3S4mur41** [reply]
Prettier just release version 3 setting `all` as [the new default for trailing comma](https://prettier.io/blog/2023/07/05/3.0.0.html#change-the-default-value-for-trailingcomma-to-all-11479httpsgithubcomprettierprettierpull11479-by-fiskerhttpsgithubcomfisker-13143httpsgithubcomprettierprettierpull13143-by-sosukesuzukihttpsgithubcomsosukesuzuki)

**Comment by MichaReiser** [comment]
## [JSX Quotes](https://prettier.io/docs/en/options.html#jsx-quotes), [Bracket Spacing](https://prettier.io/docs/en/options.html#bracket-spacing), [Bracket Line](https://prettier.io/docs/en/options.html#bracket-line), and [Arrow Function Parentheses](https://prettier.io/docs/en/options.html#arrow-function-parentheses) ##


Prettier mentions on their philosophy page that they find these options harder to motivate and only keep them because it's difficult to remove them now: 

> But other options are harder to motivate in hindsight: `--arrow-parens`, `--jsx-single-quote`, `--bracket-same-line` and `--no-bracket-spacing`are not the type of options we’re happy to have. They cause a lot of [bike-shedding](https://en.wikipedia.org/wiki/Law_of_triviality) in teams, and we’re sorry for that. Difficult to remove now, these options exist as a historical artefact and should not motivate adding more options (“If *those* options exist, why can’t this one?”).
[Source](https://prettier.io/docs/en/option-philosophy.html)]

## Proposal

Take Prettier's learning and don't add these options to Rome. Hopefully, this will help Prettier too to remove these options at some point in the future.

**Reply by ematipico** [reply]
I agree with the proposal

**Reply by vjeux** [reply]
Bracket Spacing, Bracket Line and Arrow Function Parenthesis were all added to follow Facebook internal conventions. This wasn’t the widely adopted convention outside of the company but as I was trying to get it adopted there it would have been a much bigger battle without those options.

I personally still believe that the default Facebook is using are better but that’s my opinion.

**Reply by vjeux** [reply]
For historical context about Prettier, James Long started it during the winter break but needed to go back to his main job. I convinced my manager that we needed to invest in a pretty printer for Facebook JavaScript codebase and used James Long’s few weeks old project as a base. Facebook has always been supportive of contributing to open source projects as long as it helped solve the company’s problems!

In the first 6 months I personally committed half of the commits and next 6 months a quarter. During that first year I personally reviewed almost every single pull request.

Prettier success is intimately tied with its success within Facebook that both sponsored its developement and provided (probably the best) real world use case for it.

**Reply by Conaclos** [reply]
@vjeux

> I personally still believe that the default Facebook is using are better but that’s my opinion.

Is the Facebook style guide published somewhere?

**Reply by MichaReiser** [reply]
> @vjeux
> 
> > I personally still believe that the default Facebook is using are better but that’s my opinion.
> 
> Is the Facebook style guide published somewhere?

Metro's [prettierrc](https://github.com/facebook/metro/blob/main/.prettierrc) file should give you a good idea of the code style.

**Reply by scarf005** [reply]
I suggest rome to support JSX quotes and Arrow function parenthesis.

### rationale

#### JSX single quotes
In prettierconfig using explicit settings for both `single quote` and `JSX quotes`, around 25% of them uses both single quote style. It could help consistency by using same quote for both strings and JSX.

| (both are explicit) | Single Quote | Double Quote |
| ------------------- | ------------ | ------------ |
| **JSX Single Quote**    | [6.4k][6.4k] | [251][251]   |
| **JSX Double Quote**    | [4.2k][4.2k] | [5.2k][5.2k] |

[6.4k]: https://github.com/search?type=code&q=%2F.%3FsingleQuote.%3F%3A+.%3Ftrue.%3F%2F+%2F.%3FjsxSingleQuote.%3F%3A+.%3Ftrue.%3F%2F+path%3A.prettierrc*
[251]: https://github.com/search?type=code&q=%2F.%3FsingleQuote.%3F%3A+.%3Ffalse.%3F%2F+%2F.%3FjsxSingleQuote.%3F%3A+.%3Ftrue.%3F%2F+path%3A.prettierrc*
[4.2k]: https://github.com/search?type=code&q=%2F.%3FsingleQuote.%3F%3A+.%3Ftrue.%3F%2F+%2F.%3FjsxSingleQuote.%3F%3A+.%3Ffalse.%3F%2F+path%3A.prettierrc*
[5.2k]: https://github.com/search?type=code&q=%2F.%3FsingleQuote.%3F%3A+.%3Ffalse.%3F%2F+%2F.%3FjsxSingleQuote.%3F%3A+.%3Ffalse.%3F%2F+path%3A.prettierrc*

#### Arrow function parenthesis
`23.82%` of all prettier rc explicitly uses `arrowParens: avoid`, since it has advantages addressed below:

| Options    | `avoid`        | `always`       | `always`(implicit) | total          |
| ---------- | -------------- | -------------- | ------------------ | -------------- |
| files      | [15.6k][15.6k] | [15.1k][15.1k] | 34.8k (estimated)  | [65.5k][65.5k] |
| percentage | 23.82%         | 23.05%         | 53.13% (estimated) | 100%           |

[15.6k]: https://github.com/search?type=code&q=%2F.%3FarrowParens.%3F%3A+.%3Favoid.%3F%2F+path%3A.prettierrc*
[15.1k]: https://github.com/search?type=code&q=%2F.%3FarrowParens.%3F%3A+.%3Falways.%3F%2F+path%3A.prettierrc*
[65.5k]: https://github.com/search?type=code&q=path%3A.prettierrc*

> At first glance, avoiding parentheses may look like a better choice because of less visual noise. However, when Prettier removes parentheses, it becomes harder to add type annotations, extra arguments or default values as well as making other changes. Consistent use of parentheses provides a better developer experience when editing real codebases, which justifies the default value for the option.

while this [rationale](https://prettier.io/docs/en/options.html#arrow-function-parentheses) looks reasonable for regular functions, it doesn't account for that arrow function is also used a lot as function parameters, where
1. they can be type inferred, so type hint is often redundant.
2. there are many use cases for single argument lambdas

in this typescript example:
```ts
const arr = [1, 2, 3, 4, 5]
arr.map((e: number) => e * 2).filter((e: number) => e > 3)  // (1) with explicit type hint and arrow parenthesis
arr.map((e) => e * 2).filter((e) => e > 3)                  // (2) with inferred type and arrow parenthesis
arr.map(e => e * 2).filter(e => e > 3)                      // (3) with inferred type and without arrow parenthesis
```

**Reply by scarf005** [reply]
@MichaReiser could you check above? related #3910

**Reply by vjpr** [reply]
What are the stats on `--no-bracket-spacing`?

I don't get this rule...why do we add spaces inside curly brackets, but not round or square brackets? Seems completely arbitrary. No one is typing spaces as they go.

```
foo({ bar: 1 })
foo(['bar', 'baz'])
foo('bar')
```

**Reply by jaskp** [reply]
I would actually expect `--no-bracket-spacing` to be default for Rome, as the extra space characters don't offer any objective advantage (unlike semicolons to avoid ASI bugs, arrow-parens where the extra parentheses offer easier type annotations and default values, trailing commas that can reduce the git diff when changing code, etc.)

**Reply by Conaclos** [reply]
> What are the stats on --no-bracket-spacing?

Looks like it is [under-used](https://github.com/search?type=code&q=%2FnoBracketSpacing%2F+path%3A.prettierrc*).

**Reply by jaskp** [reply]
@Conaclos it's `"bracketSpacing"`, not `"noBracketSpacing"`, so [13.7K usages](https://github.com/search?type=code&q=%2FbracketSpacing%5B%22%27%5D%3F%5Cs%3F%5B%3A%3D%5D%3F%5Cs%3Ffalse%2F+%28path%3A.prettierrc*+OR+path%3Apackage.json%29) via GitHub

**Comment by MichaReiser** [comment]
Hey @vjeux would you have a moment to share some of your experience from working at Prettier? I love Prettier's philosophy document and I have the impression you really figured it out what options you want and don't want to support! I also think it would be great if our options share the same semantics to reduce friction and confusion in the community. 

I've also been wondering what Prettier's plan is for the upcoming 3.0 release. Does Prettier plan to change any defaults or to deprecate or remove any options?

**Reply by vjeux** [reply]
Thanks for the ping! I am not longer actively involved with Prettier those days and can’t comment on the plans for Prettier 3.0. But I added historical context for all the options you mentioned.

**Reply by MichaReiser** [reply]
Thank you for these detailed explanations! My take away from your explanations are

* We should add support for `quoteProps` to properly support closure's `advanced` mode but we may not need to do so today because this mode isn't as widely used. 
* Some options were added to drive adoption at FB or by strong ask from the community. My proposal would be to not add support these options for now because Prettier standardized the code styles in the community in recent years, so the asks for these options might be lower today. This shouldn't mean that Rome should never support these options and we may want to reconsider later. 

I wasn't aware that you're no longer actively involved in Prettier's maintenance. Do you know someone who still is?

**Reply by Conaclos** [reply]
Makes sense.

**Comment by camsjams** [comment]
I would love to see the option to disable parenthesis on JSX.

Rome and Prettier rewrite code with wrapping parenthesis all over JSX components and it creates a lot of ugly noise and mess. See this largely ignored request on Prettier repo: https://github.com/prettier/prettier/issues/2724#issuecomment-1477834340

Switching to Rome doesn't seem so useful if it produces the same output as Prettier. 

Some kind of option like this on Rome would help a large list of people on that thread since Prettier won't budge on it.

**Reply by ematipico** [reply]
Yet, the introduction of **another** will create bikeshedding. Not sure if that's what you want.

**Reply by camsjams** [reply]
Prettier bikesheds about bikesheds... did you read the thread linked above and note the large amount of requests for this feature?

I admire the goal of removing options to somehow prevent bikeshedding, but this seems more like a way to baby development teams with strict rules and allow maintainers to have a reason to not add any features.
